# 01 - Data Collection

**Đề tài:** Ứng dụng Khoa học dữ liệu đa yếu tố trong phân tích, dự đoán và cảnh báo sớm nguy cơ học tập kém của sinh viên

**Bước:** 2 - Data Collection

**Mục tiêu:** Thu thập dữ liệu từ Kaggle, khảo sát cấu trúc dataset, và thiết lập lưu trữ
dữ liệu bằng MySQL để chuẩn bị cho các bước làm sạch và phân tích tiếp theo.

**Nguồn dữ liệu:** Student Performance Factors Dataset (Kaggle)

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [2]:
import pandas as pd

## 1. Đọc dữ liệu

Đọc file CSV đã tải từ Kaggle, lưu trong Google Drive.

In [3]:
df = pd.read_csv('/content/drive/MyDrive/Data science/01-Datasheet/StudentPerformanceFactors.csv')
print('Đã đọc dữ liệu thành công')

Đã đọc dữ liệu thành công


## 2. Khảo sát cấu trúc dataset

Kiểm tra số bản ghi, số thuộc tính, kiểu dữ liệu và thống kê mô tả sơ bộ của dataset.

In [4]:
print('Số dòng, số cột:', df.shape)
df.head()

Số dòng, số cột: (6607, 20)


,Hours_Studied,Attendance,Parental_Involvement,Access_to_Resources,Extracurricular_Activities,Sleep_Hours,Previous_Scores,Motivation_Level,Internet_Access,Tutoring_Sessions,Family_Income,Teacher_Quality,School_Type,Peer_Influence,Physical_Activity,Learning_Disabilities,Parental_Education_Level,Distance_from_Home,Gender,Exam_Score
0,23,84,Low,High,No,7,73,Low,Yes,0,Low,Medium,Public,Positive,3,No,High School,Near,Male,67
1,19,64,Low,Medium,No,8,59,Low,Yes,2,Medium,Medium,Public,Negative,4,No,College,Moderate,Female,61
2,24,98,Medium,Medium,Yes,7,91,Medium,Yes,2,Medium,Medium,Public,Neutral,4,No,Postgraduate,Near,Male,74
3,29,89,Low,Medium,Yes,8,98,Medium,Yes,1,Medium,Medium,Public,Negative,4,No,High School,Moderate,Male,71
4,19,92,Medium,Medium,Yes,6,65,Medium,Yes,3,Medium,High,Public,Neutral,4,No,College,Near,Female,70


In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6607 entries, 0 to 6606
Data columns (total 20 columns):
 #   Column                      Non-Null Count  Dtype 
---  ------                      --------------  ----- 
 0   Hours_Studied               6607 non-null   int64 
 1   Attendance                  6607 non-null   int64 
 2   Parental_Involvement        6607 non-null   object
 3   Access_to_Resources         6607 non-null   object
 4   Extracurricular_Activities  6607 non-null   object
 5   Sleep_Hours                 6607 non-null   int64 
 6   Previous_Scores             6607 non-null   int64 
 7   Motivation_Level            6607 non-null   object
 8   Internet_Access             6607 non-null   object
 9   Tutoring_Sessions           6607 non-null   int64 
 10  Family_Income               6607 non-null   object
 11  Teacher_Quality             6529 non-null   object
 12  School_Type                 6607 non-null   object
 13  Peer_Influence              6607 non-null   obje

In [6]:
df.describe()

,Hours_Studied,Attendance,Sleep_Hours,Previous_Scores,Tutoring_Sessions,Physical_Activity,Exam_Score
count,6607.000000,6607.000000,6607.00000,6607.000000,6607.000000,6607.000000,6607.000000
mean,19.975329,79.977448,7.02906,75.070531,1.493719,2.967610,67.235659
std,5.990594,11.547475,1.46812,14.399784,1.230570,1.031231,3.890456
min,1.000000,60.000000,4.00000,50.000000,0.000000,0.000000,55.000000
25%,16.000000,70.000000,6.00000,63.000000,1.000000,2.000000,65.000000
50%,20.000000,80.000000,7.00000,75.000000,1.000000,3.000000,67.000000
75%,24.000000,90.000000,8.00000,88.000000,2.000000,4.000000,69.000000
max,44.000000,100.000000,10.00000,100.000000,8.000000,6.000000,101.000000


## 3. Danh sách biến (Features)

Dataset thực tế gồm 19 biến đầu vào và 1 biến Exam_Score (dùng để xây dựng biến mục tiêu Risk_Level).

In [7]:
print('Số lượng cột:', len(df.columns))
df.columns.tolist()

Số lượng cột: 20


['Hours_Studied',
 'Attendance',
 'Parental_Involvement',
 'Access_to_Resources',
 'Extracurricular_Activities',
 'Sleep_Hours',
 'Previous_Scores',
 'Motivation_Level',
 'Internet_Access',
 'Tutoring_Sessions',
 'Family_Income',
 'Teacher_Quality',
 'School_Type',
 'Peer_Influence',
 'Physical_Activity',
 'Learning_Disabilities',
 'Parental_Education_Level',
 'Distance_from_Home',
 'Gender',
 'Exam_Score']

## 4. Khảo sát dữ liệu thiếu (sơ bộ)

Kiểm tra nhanh số lượng giá trị thiếu ở từng cột (việc xử lý chi tiết sẽ thực hiện ở Bước 3 - Data Cleaning).

In [8]:
df.isnull().sum()

,0
Hours_Studied,0
Attendance,0
Parental_Involvement,0
Access_to_Resources,0
Extracurricular_Activities,0
Sleep_Hours,0
Previous_Scores,0
Motivation_Level,0
Internet_Access,0
Tutoring_Sessions,0


## 5. Lưu trữ dữ liệu vào MySQL

Sử dụng dịch vụ FreeDB (MySQL online miễn phí) để lưu trữ dữ liệu, đồng thời minh họa
khả năng truy vấn SQL trên dữ liệu đã thu thập.

Cài đặt thư viện

In [9]:
!pip install pymysql sqlalchemy --quiet

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 46.7/46.7 kB 1.6 MB/s eta 0:00:00


Kết nối và đẩy dữ liệu vào MySQL

In [11]:

from sqlalchemy import create_engine

host = 'sql.freedb.tech'
database = 'freedb_7ZmRSeZ0'
username = 'u_nwODG1'
password = 'GkEGaQCdG6MH'

engine = create_engine(f'mysql+pymysql://{username}:{password}@{host}/{database}')
df.to_sql('students', engine, if_exists='replace', index=False)
print('Đã đẩy dữ liệu vào MySQL thành công')

OperationalError: (pymysql.err.OperationalError) (1045, "Access denied for user 'u_nwODG1'@'35.193.196.30' (using password: YES)")
(Background on this error at: https://sqlalche.me/e/20/e3q8)

## 6. Truy vấn dữ liệu bằng SQL

Thực hiện một số câu truy vấn SQL cơ bản để kiểm tra và khai thác dữ liệu đã lưu trữ.

In [12]:
query_result = pd.read_sql('SELECT COUNT(*) AS total_students FROM students', engine)
print(query_result)

OperationalError: (pymysql.err.OperationalError) (1045, "Access denied for user 'u_nwODG1'@'35.193.196.30' (using password: YES)")
(Background on this error at: https://sqlalche.me/e/20/e3q8)

## Kết luận

Dataset gồm 6607 bản ghi, 20 thuộc tính, đã được khảo sát cấu trúc và lưu trữ thành công
vào MySQL. Bước tiếp theo: **02_data_cleaning.ipynb** - Làm sạch dữ liệu.